RedfishPkg/RedfishHttpDxe/RedfishHttpDxe.c

```diff
@@ -0,0 +1,1336 @@
+
+  Private    = REDFISH_HTTP_CACHE_PRIVATE_FROM_THIS (This);
+  RetryCount = 0;
+  ZeroMem (Response, sizeof (REDFISH_RESPONSE));
+  ZeroMem (&Request, sizeof (REDFISH_REQUEST));
+
+  Request.Content       = Content;
+  Request.ContentLength = ContentSize;
+  Request.ContentType   = ContentType;
+
+  //
+  // Patch resource to redfish service.
+  //
+  do {
+    RetryCount += 1;
+    Status      = HttpSendReceive (
+                    Service,
+                    Uri,
+                    HttpMethodPatch,
+                    &Request,
+                    Response
+                    );
+    DEBUG ((REDFISH_HTTP_CACHE_DEBUG_REQUEST, "%a: HTTP request: %s :%r\n", __func__, Uri, Status));
+    if (!EFI_ERROR (Status) || (RetryCount >= Private->RetrySetting.MaximumRetryPatch)) {
+      break;
+    }
+
+    //
+    // Retry when BMC is not ready.
+    //
+    if ((Response->StatusCode != NULL)) {
+      DEBUG_CODE (
+        DumpRedfishResponse (NULL, DEBUG_ERROR, Response);
+        );
+
+      if (!RedfishRetryRequired (Response->StatusCode)) {
+        break;
+      }
+
+      //
+      // Release response for next round of request.
+      //
+      This->FreeResponse (This, Response);
+    }
+
+    DEBUG ((DEBUG_WARN, "%a: RedfishPatchToUriEx failed, retry (%d/%d)\n", __func__, RetryCount, Private->RetrySetting.MaximumRetryPatch));
+    if (Private->RetrySetting.RetryWait > 0) {
+      gBS->Stall (Private->RetrySetting.RetryWait);
+    }
+  } while (TRUE);
+
+  //
+  // Redfish resource is updated. Automatically expire the cached response
+  // so application can directly get resource from Redfish service again.
+  //
+  DEBUG ((REDFISH_HTTP_CACHE_DEBUG, "%a: Resource is updated, expire URI: %s\n", __func__, Uri));
+  RedfishExpireResponse (This, Uri);
+
+  if (EFI_ERROR (Status)) {
+    DEBUG_CODE (
+      DumpRedfishResponse (NULL, DEBUG_ERROR, Response);
+      );
+    //
+    // Report status code for Redfish failure
+    //
+    ReportHttpError (HttpMethodPatch, Uri, Response->StatusCode);
+    DEBUG ((DEBUG_ERROR, "%a: patch %s failed (%d/%d): %r\n", __func__, Uri, RetryCount, Private->RetrySetting.MaximumRetryPatch, Status));
+    goto ON_RELEASE;
+  }
+
+ON_RELEASE:
+
+  return Status;
+}
+
+/**
+  Perform HTTP PUT to send redfish resource to given resource URI.
+  It's caller's responsibility to free Response by calling FreeResponse ().
+
+  @param[in]  This          Pointer to EDKII_REDFISH_HTTP_PROTOCOL instance.
+  @param[in]  Service       Redfish service instance to make query.
+  @param[in]  Uri           Target resource URI.
+  @param[in]  Content       Data to put.
+  @param[in]  ContentSize   Size of the Content to be send to Redfish service.
+                            This is optional. When ContentSize is 0, ContentSize
+                            is the size of Content.
+  @param[in]  ContentType   Type of the Content to be send to Redfish service.

```

Comment:
Could we list the supported ContentType in the function header? This would be more friendly to the function users.
Same as the below occurrences.